### PR TITLE
Add support for class changing items

### PR DESCRIPTION
**Pull request by immortaleeb**
This adds support for class changing items.

As an example, I implemented the _Mysterious Card_ item which gives characters the class _Pierrot_ along with all of its abilities.

mysterious_card

pierrot_psynergies

### CODE REVIEW

File: assets/dbs/abilities_db.json [hunk 1]
```diff
@@ -2087,5 +2087,308 @@
...
+        "type": "effect_only",
+        "element": "venus",
+        "battle_target": "enemy",
+        "range": 1,
+        "pp_cost": 8,
+        "ability_power": 0,
+        "effects_outside_battle": false,
+        "is_battle_ability": true,
+        "is_field_psynergy": false,
+        "effects": [
```
Comment (jjppof) [inline assets/dbs/abilities_db.json:2309]: We need the 0.2 chance here.
image

File: base/MainChar.ts [hunk 2]
```diff
@@ -174,12 +177,23 @@ export class MainChar extends Player {
         });
     }
 
+    get granted_class_type() {
+        const equiped_class_changer = "equip_slots" in this && this.equip_slots[equip_slots.CLASS_CHANGER];
```
Comment (jjppof) [inline base/MainChar.ts:181]: no need to check `"equip_slots" in this`

File: base/MainChar.ts [hunk 3]
```diff
@@ -273,22 +287,30 @@ export class MainChar extends Player {
...
         }
+
         this.update_attributes();
         this.update_elemental_attributes();
         if (item.type === item_types.ABILITY_GRANTOR) {
             this.equipped_abilities.push(item.granted_ability);
             this.update_abilities();
         }
+
+        if (item.type === item_types.CLASS_CHANGER) {
```
Comment (jjppof) [inline base/MainChar.ts:310]: please use `else if`, so we don't need to check the `item.type` again. Also, after updating the class, we need to update the attributes, so I suggest moving `this.update_attributes();` from line 303 to just after these if elses checks.

File: base/MainChar.ts [hunk 4]
```diff
@@ -306,12 +328,18 @@ export class MainChar extends Player {
...
         this.update_elemental_attributes();
+
         if (item.type === item_types.ABILITY_GRANTOR) {
             this.equipped_abilities = this.equipped_abilities.filter(ability => {
                 return ability !== item.granted_ability;
             });
             this.update_abilities();
         }
+
+        if (item.type === item_types.CLASS_CHANGER) {
```
Comment (jjppof) [inline base/MainChar.ts:339]: Same logic from the last comment in `equip_item` function.

File: base/Classes.ts [hunk 5]
```diff
@@ -4,6 +4,7 @@ import {elements} from "./utils";
 import * as _ from "lodash";
 import {GameInfo} from "./initializers/initialize_info";
 import {Player} from "./Player";
+import {Item} from "./Item";
```
Comment (jjppof) [inline base/Classes.ts:7]: Please remove this unnecessary import.

**Review by jjppof — CHANGES_REQUESTED** [review]
Excellent work! Thanks a lot! Just change those small issues and we are ready to go.

File: base/windows/item/ClassChangeWithItemWindow.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        this.init_arrow_blinks();
+    }
+
+    private init_arrow_blinks() {
+        this.class_name_arrow_blink_timer = this.game.time.create(false);
+        this.class_name_arrow_blink_timer.loop(90, () => {
+            this.class_name_arrow.alpha = this.class_name_arrow.alpha ? 0 : 1;
+        });
+
+        this.class_name_arrow_blink_timer.start();
```
Comment (bfcferreira) [inline base/windows/item/ClassChangeWithItemWindow.ts:69]: Why do Start/Stop here and Resume down below when you can just start the timer in the "open" function?

File: base/windows/item/ItemOptionsWindow.ts [hunk 7]
```diff
@@ -311,6 +311,7 @@ export class ItemOptionsWindow {
     }
 
     open_action_message_window(text: string, close_callback: Function) {
+        console.log(text);
```
Comment (bfcferreira) [inline base/windows/item/ItemOptionsWindow.ts:314]: Console logs should never be in final commits

**Review by jjppof — APPROVED** [review]

**Pull request merged**